File: visualiztion/csv2kml.py
import numpy as np
import pandas as pd
from lxml import etree
from pykml.factory import KML_ElementMaker as KML
import pymap3d as pm
import os

'''
 mainDir - Dir where train/test data is
 data2WorkOn - Name of the data CSV file
 class2show - Which class in CSV file to draw, or (-1) if using test set or lines2show parameter
 lines2show - Which lines in CSV file to draw
 nedOrigin - Geo coordinates where XYZ = 0
 openFileWhenDone - Open KML after it is ready (KML files have to be associated with Google Earth)
'''

mainDir = 'D:\\DataHack2017\\'
#data2WorkOn = 'test'
data2WorkOn = 'train'
class2show = -1  #for 'train' only
lines2show = range(0,300) #for 'test' (or 'train if class2show == -1)
nedOrigin = (31.784491,35.214245,0)
openFileWhenDone = True


def addPlacemark(doc,name,folderName,lla_str,clr_str):
    doc.Document.Folder[folderName].append(
        KML.Placemark(
            KML.name(name),
            KML.Style(
                KML.LineStyle(
                    KML.color(clr_str),#"ff00FF40") ,
                    KML.width("4")
                )
            ),
            KML.LineString(
                KML.altitudeMode("absolute"),
                KML.coordinates(
                  lla_str
                )
            )
        )    
    )
    return doc

def ned2geodetic2String(csvSet,line):
    lla_str = ''
    for iii in range(0,15):
        northCoor = csvSet.iloc[line]['posX_'+str(iii)]
        eastCoor = csvSet.iloc[line]['posY_'+str(iii)]
        downCoor = -csvSet.iloc[line]['posZ_'+str(iii)]
        if np.isnan(northCoor):
            continue
        lla = pm.ned2geodetic(northCoor,eastCoor,downCoor,nedOrigin[0],nedOrigin[1],nedOrigin[2])
        lla_str = lla_str+str(lla[1])+","+str(lla[0])+","+str(lla[2])+" "
    return lla_str
    
name = 'Rocket_Data_Science'
if class2show > 0:
    name = name+'_'+str(class2show)
doc = KML.kml(
        KML.Document(
            KML.name(name),
            KML.Folder(
                KML.name('UP')
            ),
            KML.Folder(
                KML.name('DOWN')
            )
            
        )
    )
    
print 'Loading...'
csvSet = pd.read_csv(mainDir+data2WorkOn  + '.csv')

print 'Working...'
if data2WorkOn.find('train') and class2show > 0:
    csvSubSet = csvSet[csvSet.iloc[:]['class'] == class2show]
    lines2show = csvSubSet.index
for line in lines2show:
    lla_str = ned2geodetic2String(csvSet,line)
    velColor = hex(int(round(np.linalg.norm( ( csvSet.iloc[line]['velX_0']  ,  csvSet.iloc[line]['velY_0'] , csvSet.iloc[line]['velZ_0']  ) ) / 1500 * 255)))[-2:]    
    if csvSet.iloc[line]['velZ_0'] > 0:    
        clr_str = "FF"+velColor+"0000"
        folderName = 0
    else:
        clr_str = "FF0000"+velColor
        folderName = 1
    addPlacemark(doc,'line_'+str(line),folderName,lla_str,clr_str)

print 'Saving...'
#print etree.tostring(doc, pretty_print=True)
outfile = file(mainDir+name+'.kml','w')
outfile.write(etree.tostring(doc, pretty_print=True))

print 'Done!'
if openFileWhenDone:
    os.startfile(mainDir+name+'.kml')

File: visualiztion/csv2kml.m
function csv2kml(mainDir,data2WorkOn,class2show,lines2show,nedOrigin,openFileWhenDone)
%%%%
% mainDir - Dir where train/test data is
% data2WorkOn - Name of the data CSV file
% class2show - Which class in CSV file to draw, or (-1) if using test set or lines2show parameter
% lines2show - Which lines in CSV file to draw
% nedOrigin - Geo coordinates where XYZ = 0
% openFileWhenDone - Open KML after it is ready (KML files have to be associated with Google Earth)
%%%%
if nargin == 0
	mainDir = 'D:\DataHack2017\';
	%data2WorkOn = 'test';
	data2WorkOn = 'train';	
	class2show = -1;  %for 'train' only
	lines2show = 1:300; %for 'test' (or 'train if class2show == -1)
	nedOrigin = [31.784491,35.214245,0];
	openFileWhenDone = 1;
end



name = 'Rocket_Data_Science';
if class2show > 0
    name = [name '_' num2str(class2show)];
end
kmlStrHeader = ['<kml xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:atom="http://www.w3.org/2005/Atom" xmlns="http://www.opengis.net/kml/2.2"> \n' ...
'  <Document> \n' ...
'    <name>' name '</name> \n'];

kmlStrFooter = '  </Document></kml>';
strUP = '';
strDN = '';

disp('Loading...');
csvSetTable = readtable([mainDir data2WorkOn '.csv']);
csvSet = table2array(csvSetTable(:,1:211));

disp('Working...');
if strfind(data2WorkOn,'train') && class2show > 0
    lines2show = find(csvSet(:,end) == class2show);
end

for lll = lines2show(:)'
    lla_str = ned2geodetic2String(csvSet,lll,nedOrigin);
    velVal = round((norm( csvSet(lll,6:8) )/1500) * 255 );
    velColor = dec2hex(velVal,2);
    if  csvSet(lll,8) > 0
      clr_str = ['FF' velColor '0000'];
      strUP = addPlacemark(strUP,['line_' num2str(lll-1)],lla_str,clr_str);
    else
      clr_str = ['FF0000' velColor];
      strDN = addPlacemark(strDN,['line_' num2str(lll-1)],lla_str,clr_str);      
    end
end

finalStr = [kmlStrHeader '<Folder>\n <name>UP</name> \n' strUP ...
    '</Folder><Folder>\n <name>DOWN</name> \n' strDN '</Folder>' kmlStrFooter];

disp('Saving...');
fid = fopen([mainDir name '_MATLAB.kml'],'w');
fprintf(fid,finalStr);
fclose(fid);

disp('Done!');
if openFileWhenDone
   winopen( [mainDir name '_MATLAB.kml']);
end

end

function str = addPlacemark(str,name,lla_str,clr_str)

strTMP = [ ...
'    <Placemark> \n' ...
'      <name>' name '</name>  \n' ...
'      <Style>  \n' ...
'        <LineStyle>  \n' ...
'          <color>' clr_str '</color> \n' ...
'          <width>4</width> \n' ...
'        </LineStyle>  \n' ...
'      </Style> \n' ...
'      <LineString> \n' ...
'        <altitudeMode>absolute</altitudeMode> \n' ...
'        <coordinates>'  lla_str '</coordinates> \n' ...
'      </LineString> \n' ...
'    </Placemark>     \n' ...
];

str = [str strTMP];

end


function lla_str = ned2geodetic2String(csvSet,line,nedOrigin)
lla_str = '';
for iii = 1:15
    northCoor = csvSet(line,3+7*(iii-1));
    eastCoor = csvSet(line,4+7*(iii-1));
    downCoor = -csvSet(line,5+7*(iii-1));
    if isnan(northCoor)
       continue 
    end
    [lat,lon,h] = ned2geodetic(northCoor,eastCoor,downCoor,nedOrigin(1),nedOrigin(2),nedOrigin(3),referenceEllipsoid('GRS 80'));
    lla_str = [lla_str num2str([lon lat h],'%.10f,%.10f,%.10f') ' '];
end

end
